### DISCUSSION TITLE
Is ISR and dynamic tenants supported?

### DISCUSSION THREAD

**Discussion by nicosh** [micheleriva/krabs#4] [Q&A] [answered]
Hi,
First of all thanks for krabs it looks awesome, bravo!
I have just a couple of questions , my use case is this : 
I have a server which points any hostname to next.js+krabs app, this app will handle all the requests with  a catch all route : `pages/[...path].js`, so i wouldn't have any `.krabs.js` config file, instead  i would like to have krabs to automatically point any reques to that route. `[...path].js` will be able to know from which hostname the request come from?

So is [ISR](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) with fallback supported?
Can we use Krabs without specifying number of tenenats?   
Thanks.

**Comment by micheleriva** [comment, marked as answer]
Hi @nicosh!
With Next.js, you can always have access to the `hostname` variable as it is a property of a standard Express.js' `req` object. If you're using a custom server, you can also apply custom middleware to elaborate `hostname` and adding some more information to the `req` object, which will be sent to Next.js later.

- From what I tested, ISR is still supported
- You can use dynamic tenants as the Krabs config is just a JavaScript file, and it can depend on external data sources such as APIs

Does this answer your questions?
Cheers!

**Reply by nicosh** [reply]
@micheleriva  Thanks,
>You can use dynamic tenants as the Krabs config is just a JavaScript file  

But in this case config  will be an async function or a function that will return a promise, still supported?
In case that i dont have a list of know hostname there is no way to do somethink like this? 
```
module.exports = {
  tenants: [
    {
      name: '*', // something like this, any name
      domains: [
        {
          dev: /dev\.[a-z]*\.local\.website-1\.com/, // Regex supported!
          stage: 'stage.website-1.com',
          prod: 'website-1.com',
        },
      ],
    }
  ],
};
```
Thanks

**Reply by nicosh** [reply]
@micheleriva also this is not really true 
>...more information to the req object, which will be sent to Next.js later.  

Sadly with `getStaticProps` we can only access to `path`, that will not contain hostname, (unless it is clearly exposed in the url, but it make no sense then).
Also even if i know all the tenants (fetching them inside config file) then the only way to handle all of them is to create a catch all route, but then this will not work with ISR
For reproduce : 
Create a catch all router ex. `pages/[[fullpath].js` 
with getStaticPaths
```
export async function getStaticPaths() {
  return { paths: [], fallback: true };
}
```
and getStaticProps  
```
export const getStaticProps = async (context) => {
  console.log(context) // no hostname
  let data = await loadPage(); //  here we should load data for hostname+path
  return {
    props: {
      json:  data
    },
    revalidate: 1, // In seconds
  }
}
```

**Reply by micheleriva** [reply]
Oh, I see your point. Let me do some tests and I'll reach back!

**Reply by GodGardensGuns** [reply]
In your [youtube video](https://www.youtube.com/watch?v=71NRAnT4G4Q) at around 15:10, you mention that ISR is not yet supported. Just wanted to confirm one way or the other.

**Reply by micheleriva** [reply]
@dayvista let me paste my comment from the other discussion:

> If you really need SSG + ISR, there's a quick solution for that: you can use a CDN as a caching layer. That way, Cloudflare, Akamai, Fastly, or any other CDN will cache your pages for x minutes and serve the cached version until a given TTL allows it. Does it make sense? 🙂

That way, you can make your website support SSG + ISR... without having to support SSG + ISR. By the way, I'm still working for those who don't want to use a CDN for any reason!

**Reply by micheleriva** [reply]
Hi @nicosh !

> But in this case config will be an async function or a function that will return a promise, still supported?
In case that i dont have a list of know hostname there is no way to do somethink like this?

Just to follow-up, with the newest 0.0.6 version, Krabs config supports both async functions and objects!
